tensorflow/core/kernels/conv_ops_gpu_3.cu.cc

```diff
@@ -413,6 +429,275 @@ struct PadInput<GPUDevice, T, int, NDIMS> {
+                                       total_tiles_count, input, input_dims,
+                                       output);
+    }
+  }
+};
+
+template <typename T, int TileLongSide, int TileShortSide>
+struct BatchNarrowMatrixTransposeDispatcher<
+    T, TileLongSide, TileShortSide,
+    typename std::enable_if<TileSizePossibilityFrontierCheck(
+                                TileLongSide, TileShortSide, sizeof(T), 0),
+                            void>::type> {
+  static void DoIt(const GPUDevice& d, int tile_size_i,
+                                           int tile_size_j,
+                                           int total_tiles_count,
+                                           const T* input,
+                                           const Dimension<3>& input_dims,
+                                           T* output) {
+
+    static_assert((TileLongSide & (TileLongSide - 1)) == 0,
+      "The length of the longer side of the tile is always a power of 2.");
+
+    const int NumThread = TileLongSide;
+    if (tile_size_i <= TileLongSide && tile_size_j <= TileShortSide)
+      SwapDimension1And2InTensor3UsingTiles<
+          T, NumThread, TileLongSide,
+          TileShortSide><<<total_tiles_count, NumThread, 0, d.stream()>>>(
+          input, input_dims, output);
+    else if (tile_size_j <= TileLongSide && tile_size_i <= TileShortSide)
+      SwapDimension1And2InTensor3UsingTiles<
+          T, NumThread, TileShortSide,
+          TileLongSide><<<total_tiles_count, NumThread, 0, d.stream()>>>(
+          input, input_dims, output);
+    return;
+  }
+};
+
+template <typename T, int TileLongSide, int TileShortSide>
+struct BatchNarrowMatrixTransposeDispatcher<
+    T, TileLongSide, TileShortSide,
+    typename std::enable_if<TileSizePossibilityFrontierCheck(
+                                TileLongSide, TileShortSide, sizeof(T),
+                                /* mode = outside */ 1),
+                            void>::type> {
+  static void DoIt(const GPUDevice& d, int tile_size_i,
+                                           int tile_size_j,
+                                           int total_tiles_count,
+                                           const T* input,
+                                           const Dimension<3>& input_dims,
+                                           T* output) {
+    assert(false &&
+           "BatchNarrowMatrixTransposeDispatcher has requested an unexpected "
+           "launch configuration. ");
+  }
+};
+
+// A helper function to make RunSwapDimension1And2InTensor3 concise. This
+// helper function looks at the data type and input matrix sizes and decides
+// the thread numbers and tile sizes to use.
+template <typename T>
+void SwapDimension1And2InTensor3WithNarrowMatrices(
+    const GPUDevice& d, const T* input, const Dimension<3>& input_dims,
+    T* output, const int kMinDimensionToUseTiles) {
+  // Define available tile sizes here for each size of data type supported:
```

Comment:
If this needs to be kept in sync with something else, we should probably say so...

In fact it would be really nice if we could autogenerate it from the frontier function?  It seems like this might actually be feasible, since the search space is small.  One only needs to compute it once; you can cache the result in a function-local static variable.

The trick is that the function-local static variable must be a POD, according to our style rules.  So a vector is not OK, but a vector* is fine.  For example:

```
// (Some meaningful comment...)
template <int SizeOfT>
const std::vector<std::pair<int, int>>& GetTileSizes() {
  static_assert(SizeOfT <= 16);
  static auto* sizes = [] {
    auto* sizes = new std::vector<pair<int, int>>();
    // Expensive work to populate sizes, lazily run in a thread-safe manner the first time GetTileSizes<N> is called.
    return sizes;
  }();
  return *sizes;
}
```